### PR TITLE
Fixing invalid cast when autoproperty is initialized in another initi…

### PR DESCRIPTION
**Pull request by VSadov**
…alizer.

Watson crash 1175559

### CODE REVIEW

**Comment by VSadov** [comment]
* Bug number - fixes Watson crash 1175559  
* Customer scenario - autoproperty is initialized in another initializer results in an invalid cast crash. This is rarely valid, since in C# initializers cannot access "this", but crashing is still bad.
* The scenario is more common in VB and VB does not have the bug due to slight differences in implementation here.
* Fix description - do not just assume that containing member is always a method on this codepath, check for that.
* Testing done - new unit-tests scenarios.
@jaredpar FYI

**Comment by VSadov** [comment]
@gafter @AlekseyTs @agocke @khyperia - please take a look.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 1]
```diff
@@ -1959,11 +1959,17 @@ internal static bool AccessingAutopropertyFromConstructor(BoundExpression receiv
...
                     sourceProperty.IsAutoProperty &&
                     sourceProperty.ContainingType == fromMember.ContainingType &&
-                    ((MethodSymbol)fromMember).MethodKind == (propertyIsStatic ? MethodKind.StaticConstructor
-                                                                                : MethodKind.Constructor) &&
-                   (propertyIsStatic || receiver.Kind == BoundKind.ThisReference);
+                    IsConstructorOrField(fromMember, isStatic: propertyIsStatic) &&
+                    (propertyIsStatic || receiver.Kind == BoundKind.ThisReference);
         }
 
+        private static bool IsConstructorOrField(Symbol member, bool isStatic)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: This function doesn't quite do what it says -- it doesn't ask whether or not the member is a constructor, it asks if it's a constructor with the same receiver type as given in isStatic. I would consider changing the name of the method or the parameter name to make it clear that the parameter is meant to indicate the expected receiver of the member.

**Comment by agocke** [comment]
LGTM, except for given comment.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by jaredpar** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 1 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: This is basically a part of a complicated condition refactored into a private helper and is used in a single place (could be a good case for nested methods). 
I am not able to name it much better :-). I can consider any suggestions, but as far as names I can come up with, the current one seems ok.

**Comment by MattGertz** [comment]
Approved 5/27 by ML Shiproom.

**Pull request merged**